gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,251 @@
+        if client:
+
+            def notify_client_active(*args):
+                if not self.native.get_client().props.active:
+                    # If the client explicitly becomes inactive, this indicates
+                    # a failure to retrieve the location
+                    # e.g., when the geoclue service is stopped on the host
+                    self.props.state = State.FAILED
+                    self._stop_tracking()
+                else:  # pragma: no cover
+                    # Not possible to meaningfully test in a platform agnostic manner
+                    self.props.state = State.READY
+
+            self.notify_active_listener = client.connect(
+                "notify::active", notify_client_active
+            )
+
+    def has_permission(self):
+        return bool(self.permission_result)
+
+    def request_permission(self, result):
+        if self.permission_result is not None:
+            result.set_result(self.permission_result)
+            return
+
+        def wait_for_client(*args):
+            if State.is_uninitialised(self.props.state):  # pragma: no cover
+                return
+
+            self.permission_result = State.is_available(self.props.state)
+
+            result.set_result(self.permission_result)
+            self.disconnect(listener_handle)
+
+        listener_handle = self.connect("notify::state", wait_for_client)
+
+        if self.props.state == State.INITIAL:
+            self._start()
+        else:  # pragma: no cover
+            # This cannot be tested because the test probe runs a sync initialisation
+            # There is no method by which initialisation can be triggered in the test
+            # environment and then permissions requested before initialisation finishes
+            # In other words, there is no situation in test where the status is starting
+            pass
+
+    def has_background_permission(self):
+        """Check for background permission.
+
+        Background location permission has no meaning for Geoclue,
+        as all location access is mediated through the same location
+        tracking APIs. Therefore, background location permission
+        is handled in identical terms to foreground location permission.
+        """
+        return self.has_permission()
+
+    def request_background_permission(self, result):
+        """Request background permission.
+
+        This method should never actually run, because the upstream location
+        API enforces requesting general location permissions first. Once
+        general location permissions are available, background permission
+        will already be available (see :meth:`~.Location.has_background_permission()`),
+        and the upstream location API never has to call this method.
+        """
+        raise NotImplementedError("Background permissions are non-existent on Linux")
```

Comment:
Sure thing. It'll have to have a `no cover` pragma on it, though, because `has_background_permission` is just a call to `has_permission` in this implementation. When the Toga user calls `request_background_permission`, the core `Location` class will bypass the actual request because `has_background_permission` will already be true (and must be, because you have to request foreground permission before background). In other words, there's no way to get this method to run via the core `Location` api. I can add that in a comment explaining why it can't be covered, though.

Reply:
Sure - as long as the code has a brief explanatory note for why the no-cover is justified, that's fine.

Reply:
Done in https://github.com/beeware/toga/pull/2999/commits/280065a02a5c13fcfec1d97bea4360bee3f12da4